File: solutions/python/bank-account/1/bank_account.py
import uuid

class BankAccount:
    def __init__(self):
        self.active = False
        self.saldo = 0
        self.id = str(uuid.uuid4())


    def get_balance(self):
        if self.active:
            return self.saldo
        else:
            raise ValueError('account not open')

    def open(self):
        if not self.active:
            self.active = True
        else:
            raise ValueError('account already open')

    def deposit(self, amount):
        if not self.active:
            raise ValueError('account not open')
        if amount <= 0:
            raise ValueError('amount must be greater than 0')
        else:
            self.saldo += amount

    def withdraw(self, amount):
        if not self.active:
            raise ValueError('account not open')
        if amount <= 0:
            raise ValueError('amount must be greater than 0')        
        if self.saldo < amount:
            raise ValueError('amount must be less than balance')
        else:
            self.saldo -= amount

    def close(self):
        if self.active:
            self.active = False
            self.saldo = 0
        else:
            raise ValueError('account not open')

File: solutions/javascript/bank-account/1/bank-account.js
//
// This is only a SKELETON file for the 'Bank Account' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class BankAccount {
  constructor() {
    this.status = false;
    this.saldo = 0;
  }

  open() {
    if (!this.status) this.status = true;
    else throw new ValueError();
  }

  close() {
    if (this.status) {
      this.status = false;
      this.saldo = 0;
    } else throw new ValueError();
  }

  deposit(money) {
    if (this.status && money > 0) this.saldo += money;
    else throw new ValueError();
  }

  withdraw(money) {
    if (this.status && money > 0 && money <= this.saldo) this.saldo -= money;
    else throw new ValueError();
  }

  get balance() {
    if (this.status) return this.saldo;
    else throw new ValueError();
  }
}

export class ValueError extends Error {
  constructor() {
    super("Bank account error");
  }
}
